package com.airtibe.meditrack.service;

import com.airtibe.meditrack.entity.Appointment;
import com.airtibe.meditrack.entity.Bill;
import com.airtibe.meditrack.entity.BillSummary;
import com.airtibe.meditrack.entity.PaymentReceipt;
import com.airtibe.meditrack.enums.AppointmentStatus;
import com.airtibe.meditrack.repository.BillRepository;
import com.airtibe.meditrack.enums.PaymentStatus;


public class BillingService {

    private final BillRepository repository;
    private final AppointmentService appointmentService;

    public BillingService(
            BillRepository repository,
            AppointmentService appointmentService) {

        this.repository = repository;
        this.appointmentService = appointmentService;
    }

    public BillSummary finalizeBill(String appointmentId) {

        Appointment appointment = appointmentService.searchById(appointmentId);

        if (appointment == null) {
            return null;
        }

        Bill bill = new Bill(appointment);

        bill.addItem("Consultation Fee", 500);
        bill.addItem("Medical Service", 300);

        BillSummary summary = bill.generateBillSummary();

        repository.save(bill);

        appointment.setStatus(AppointmentStatus.BILL_FINALIZED);

        return summary;
    }

    public PaymentReceipt proceedToPayment(BillSummary billSummary) {

        PaymentReceipt receipt = new PaymentReceipt(
                billSummary.getBillSummaryId(),
                billSummary.getTotalAmount(),
                "UPI",
                PaymentStatus.SUCCESS
        );

        return receipt;
    }
}